friends/src/components/AddFriendForm.js

```diff
@@ -0,0 +1,80 @@
+import React from "react";
```

Comment:
This file was tricky. We really need to be using fully controlled inputs, which means we need to have each value controlled by state. I know it works without value defined, but it goes against react convention. Also I found your conditions (like in input default value, or inside put friend) to be confusing and hard to read. My proposed solution would be to use react lifecycle methods to check the selected id, find to selected friend, and populate your existing state with those values. Then, you can just set your input values to this.state.name, you don't have to keep checking for selectedFriend. The trick here is you will also need to use componentDidUpdate for when the selectedID is updated. Also, where is this.initialState being defined? I can't find it, and I don't see any of your state being reset. I would define initial state as a const outside the class. Once you have your state holding the relevant values, refactor your update and add methods to only refer to the state. Also, if you have both the add and update showing at the same time, how can you be sure the user will click the right one? I will share a snippet of my proposed changes a little ways below here, I know you'll want to try it yourself first, but I'm sure my instructions here aren't 100% clear, and I'd like you to see what I am talking about .





```js
import React from "react";

const initialState = {
  name: "",
  age: "",
  email: "",
  isUpdating: false
};

export default class AddFriendForm extends React.Component {
  constructor() {
    super();
    this.state = {
      name: "",
      age: "",
      email: "",
      isUpdating: false
    };
  }
  // Rather than having to check selected friend for everything we need, let's just use our existing state, and set selected friend's properties to it. This way our inputs can be fully controlled with value. Our inputs will always pull their value from this.state.name/age/email. ComponentDidMount populates the state with the selected friend if it exists, and componentDidUpdate will make sure it updates when the component receives new props. Honestly lines 14-21 and 25-32 could maybe be extracted into a function too. We also need the isUpdating state for when we get to the render method
  componentDidMount() {
    const selectedFriend = this.props.friendsList.find(
      friend => friend.id === this.props.selectedID
    );
    this.setState({
      name: selectedFriend ? selectedFriend.name : "",
      age: selectedFriend ? selectedFriend.age : "",
      email: selectedFriend ? selectedFriend.email : "",
      isUpdating: selectedFriend ? true : false
    });
  }
  componentDidUpdate(prevProps) {
    if (this.props.selectedID !== prevProps.selectedID) {
      const selectedFriend = this.props.friendsList.find(
        friend => friend.id === this.props.selectedID
      );
      this.setState({
        name: selectedFriend ? selectedFriend.name : "",
        age: selectedFriend ? selectedFriend.age : "",
        email: selectedFriend ? selectedFriend.email : "",
        isUpdating: selectedFriend ? true : false
      });
    }
  }

  handleChanges = e => {
    this.setState({
      [e.target.name]: e.target.value
    });
  };

  addFriend = e => {
    e.preventDefault();
    this.props.postFriend({
      name: this.state.name,
      age: Number(this.state.age),
      email: this.state.email
    });
    // Where is initial state being defined?
    this.setState(() => initialState);
  };

  updateFriend = e => {
    e.preventDefault();
    this.props.putFriend({
      name: this.state.name,
      age: Number(this.state.age),
      email: this.state.email
    });
    this.setState(() => initialState);
  };

  render() {
    return (
      <form
        onSubmit={this.state.isUpdating ? this.updateFriend : this.addFriend}
      >
        <input
          placeholder="Name"
          onChange={this.handleChanges}
          value={this.state.name}
          name="name"
          required
        />
        <input
          placeholder="Age"
          onChange={this.handleChanges}
          value={this.state.age}
          name="age"
          type="number"
        />
        <input
          placeholder="Email"
          onChange={this.handleChanges}
          value={this.state.email}
          name="email"
          type="email"
        />
        <button>
          {this.state.isUpdating ? "Update Friend" : "Add Friend"}
        </button>
      </form>
    );
  }
}
```